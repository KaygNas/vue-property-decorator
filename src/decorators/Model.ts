import {
  createDecorator,
  PropOptions,
  Vue,
  VueDecorator,
} from 'vue-class-component'

type Constructor = (new () => any) | SymbolConstructor

/**
 * Decorator for v-model
 * @param propName e.g. `modelValue`
 * @param propOptions the options for the prop
 */
export function Model(
  propName: string,
  propOptions?: Constructor | Constructor[] | PropOptions,
): VueDecorator {
  return createDecorator((componentOptions, key) => {
    const eventName = `update:${propName}`
    componentOptions.props ||= Object.create(null)
    componentOptions.props[propName] = propOptions
    componentOptions.emits ||= []
    componentOptions.emits.push(eventName)
    componentOptions.computed ||= Object.create(null)
    componentOptions.computed[key] = {
      get(this: any) {
        return this[propName]
      },
      set(this: Vue, newValue: any) {
        this.$emit(eventName, newValue)
      },
    }
  })
}

/**
 * @deprecated
 *
 * Decorator for v-model having same behaviour in Vue2
 * @param event will be ignore and always be 'update:modelValue' as vue3 specify.
 * @param propOptions the options for the prop
 */
export function LegacyModel(
  inputEventName: string,
  propOptions?: Constructor | Constructor[] | PropOptions,
) {
  const propName = 'modelValue';
  const eventName = `update:${propName}`
  return createDecorator((componentOptions, key) => {
    componentOptions.props ||= Object.create(null)
    componentOptions.props[propName] = propOptions

    componentOptions.computed ||= Object.create(null)
    componentOptions.computed[key] = {
      get(this: any) {
        return this[propName]
      },
      set(this: Vue, newValue: any) {
        this.$emit(eventName, newValue)
      },
    }

    let self: Vue
    componentOptions.emits = Array.isArray(componentOptions.emits) 
    ? Object.fromEntries(componentOptions.emits.map(emit => [emit, null])) 
    : { [eventName]: null }
    const created = componentOptions.created 
    componentOptions.created = function (this: any) {
      created?.call(this)
      self = this
    }
    // intercept $emit, change inputEventName into update:modelValue
    componentOptions.emits[inputEventName] = function (this:any, ...payload:any) {
      self.$emit(eventName, ...payload)
      return true
    }
  })
}
