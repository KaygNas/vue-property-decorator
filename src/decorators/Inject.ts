import { inject, InjectionKey } from 'vue'
import { createDecorator, VueDecorator } from 'vue-class-component'

export type InjectOptions = {
  from?: string | InjectionKey<any>
  default?: any
}

/**
 * Decorator for inject options
 * @param options the options for the injected value
 */
export function Inject(
  options: InjectOptions = Object.create(null),
): VueDecorator {
  return createDecorator((componentOptions, key) => {
    const originalSetup = componentOptions.setup
    componentOptions.setup = (props, ctx) => {
      const result = originalSetup?.(props, ctx)
      const injectedValue = inject(options.from || key, options.default)
      return {
        ...result,
        [key]: injectedValue,
      }
    }
  })
}

/**
 * @deprecated - use Inject({from: 'key', default: 'default'}) instead
 * Decorator for legacy InjectReactive
 */
export function LegacyInjectReactive(
  keyOrOptions: InjectOptions | InjectOptions['from'],
) {
  if (
    typeof keyOrOptions === 'string' ||
    typeof keyOrOptions === 'undefined' ||
    typeof keyOrOptions === 'symbol' ||
    keyOrOptions instanceof Symbol
  ) {
    return Inject({ from: keyOrOptions })
  } else {
    return Inject(keyOrOptions)
  }
}
