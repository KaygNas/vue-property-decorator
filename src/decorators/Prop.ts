import { createDecorator, PropOptions, VueDecorator } from 'vue-class-component'

type Constructor = (new () => any) | SymbolConstructor

/**
 * Decorator for prop options
 * @param propOptions the options for the prop
 */
export function Prop(
  propOptions?: Constructor | Constructor[] | PropOptions,
): VueDecorator {
  return createDecorator((componentOptions, key) => {
    if (typeof componentOptions.props === 'undefined') {
      componentOptions.props = Object.create(null)
    }
    componentOptions.props[key] = propOptions
  })
}
