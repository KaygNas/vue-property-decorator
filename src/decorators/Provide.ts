import { computed } from 'vue'
import { createDecorator, VueDecorator } from 'vue-class-component'

/**
 * Decorator for provide options
 * @param to to
 */
export function Provide(to?: string): VueDecorator {
  return createDecorator((componentOptions, key) => {
    const originalProvide = componentOptions.provide
    componentOptions.provide = function (this: any) {
      const providedValue =
        typeof originalProvide === 'function'
          ? originalProvide.call(this)
          : originalProvide

      return {
        ...providedValue,
        [to || key]: computed(() => this[key]),
      }
    }
  })
}
