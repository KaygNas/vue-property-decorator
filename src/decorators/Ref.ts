import { createDecorator, Vue, VueDecorator } from 'vue-class-component'

/**
 * decorator of a ref prop
 * @param refKey the ref key defined in template
 */
export function Ref(refKey?: string): VueDecorator {
  return createDecorator((componentOptions, key) => {
    // some hacky trick to force componentOptions.computed[key]
    // compute every time component updated
    const refSymbol = {
      symbol: Symbol(refKey || key),
      get: (vm: any) => vm[refSymbol.symbol],
      set: (vm: any) => vm[refSymbol.symbol] = Symbol(refKey || key)
    }

    const createData = componentOptions.data ?? (() => ({}))
    componentOptions.data = function () {
      const data = createData.call(this, this)
      refSymbol.set(data)
      return data
    }

    const created = componentOptions.created
    componentOptions.created = function () {
      created?.call(this)
      refSymbol.get(this)
    }

    const beforeUpdate = componentOptions.beforeUpdate
    componentOptions.beforeUpdate = function () {
      beforeUpdate?.call(this)
      refSymbol.set(this)
    }

    componentOptions.computed ||= Object.create(null)
    componentOptions.computed[key] = {
      cache: false,
      get(this: Vue) {
        refSymbol.get(this)
        return this.$refs[refKey || key]
      },
    }
  })
}
