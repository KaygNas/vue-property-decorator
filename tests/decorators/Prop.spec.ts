import { mount, VueWrapper } from '@vue/test-utils'
import { h } from 'vue'
import { Vue } from 'vue-class-component'
import { Prop } from '../../src/decorators/Prop'

describe(Prop, () => {
  describe('without argument', () => {
    class MyComponent extends Vue {
      @Prop() stringProp: string | undefined
      @Prop() numberProp: number | undefined
      @Prop() booleanProp: boolean | undefined
      @Prop() arrayProp: Array<any> | undefined
      @Prop() objectProp: Record<string, any> | undefined
      @Prop() dateProp: Date | undefined
      @Prop() functionProp: Function | undefined
      @Prop() symbolProp: Symbol | undefined

      render() {
        return h('div')
      }
    }

    let wrapper: VueWrapper<MyComponent>

    describe('without prop values from its parent component', () => {
      beforeEach(() => {
        wrapper = mount(MyComponent)
      })

      it('sets props correctly', () => {
        expect(wrapper.vm.$options.props).toMatchSnapshot()
      })

      const properties: (keyof MyComponent)[] = [
        'stringProp',
        'numberProp',
        'booleanProp',
        'arrayProp',
        'objectProp',
        'functionProp',
        'symbolProp',
      ]

      it.each(properties)(`%s is undefined`, (property) => {
        expect(wrapper.vm[property]).toBeUndefined()
      })
    })

    describe('with prop values from its parent component', () => {
      beforeEach(() => {
        wrapper = mount(MyComponent, {
          props: {
            stringProp: 'STRING-PROP',
            numberProp: 2000,
            booleanProp: true,
            arrayProp: [],
            objectProp: {},
            dateProp: new Date('2020-01-01'),
            functionProp: new Function(),
            symbolProp: Symbol.for('SYMBOL-PROP'),
          },
        })
      })

      const properties: [keyof MyComponent, any][] = [
        ['stringProp', 'STRING-PROP'],
        ['numberProp', 2000],
        ['booleanProp', true],
        ['arrayProp', []],
        ['objectProp', {}],
        ['dateProp', new Date('2020-01-01')],
        ['functionProp', expect.any(Function)],
        ['symbolProp', Symbol.for('SYMBOL-PROP')],
      ]

      it.each(properties)(
        `%s equals to the given prop value`,
        (property, expected) => {
          expect(wrapper.vm[property]).toEqual(expected)
        },
      )
    })
  })

  describe('with constructor', () => {
    class MyComponent extends Vue {
      @Prop(String) stringProp: string | undefined
      @Prop(Number) numberProp: Number | undefined
      @Prop(Boolean) booleanProp!: boolean
      @Prop(Array) arrayProp: Array<any> | undefined
      @Prop(Object) objectProp: Record<string, any> | undefined
      @Prop(Date) dateProp: Date | undefined
      @Prop(Function) functionProp: Function | undefined
      @Prop(Symbol) symbolProp: Symbol | undefined

      render() {
        return h('div')
      }
    }

    let wrapper: VueWrapper<MyComponent>

    describe('without prop values from its parent component', () => {
      beforeEach(() => {
        wrapper = mount(MyComponent)
      })

      it('sets props correctly', () => {
        expect(wrapper.vm.$options.props).toMatchSnapshot()
      })

      const properties: [keyof MyComponent, any][] = [
        ['stringProp', undefined],
        ['numberProp', undefined],
        ['booleanProp', false],
        ['arrayProp', undefined],
        ['objectProp', undefined],
        ['dateProp', undefined],
        ['functionProp', undefined],
        ['symbolProp', undefined],
      ]

      it.each(properties)('%s equals to %s', (property, expected) => {
        expect(wrapper.vm[property]).toEqual(expected)
      })
    })

    describe('with prop values from its parent component', () => {
      beforeEach(() => {
        wrapper = mount(MyComponent, {
          props: {
            stringProp: 'STRING-PROP',
            numberProp: 300,
            booleanProp: true,
            arrayProp: [],
            objectProp: {},
            dateProp: new Date('2020-01-01'),
            functionProp: new Function(),
            symbolProp: Symbol.for('SYMBOL-PROP'),
          },
        })
      })

      const properties: [keyof MyComponent, any][] = [
        ['stringProp', 'STRING-PROP'],
        ['numberProp', 300],
        ['booleanProp', true],
        ['arrayProp', []],
        ['objectProp', {}],
        ['dateProp', new Date('2020-01-01')],
        ['functionProp', expect.any(Function)],
        ['symbolProp', Symbol.for('SYMBOL-PROP')],
      ]

      it.each(properties)(
        `%s equals to the given prop value`,
        (property, expected) => {
          expect(wrapper.vm[property]).toEqual(expected)
        },
      )
    })
  })

  describe('with multiple constructors', () => {
    class MyComponent extends Vue {
      @Prop([String, Number]) myId!: string | number

      render() {
        return h('div')
      }
    }

    let wrapper: VueWrapper<MyComponent>

    beforeEach(() => {
      wrapper = mount(MyComponent)
    })

    it('sets prop options correctly', () => {
      expect(wrapper.vm.$options.props).toMatchSnapshot()
    })
  })

  describe('with prop option', () => {
    class MyComponent extends Vue {
      @Prop({ type: String, default: 'DEFAULT-STRING' }) stringProp!: string
      @Prop({ type: Number, default: 1234 }) numberProp!: number
      @Prop({ type: Boolean, default: true }) booleanProp!: boolean
      @Prop({ type: Array, default: () => ['DEFAULT-ARRAY'] })
      arrayProp!: Array<any>
      @Prop({ type: Object, default: () => ({ value: 'DEFAULT-OBJECT' }) })
      objectProp!: Record<string, any>
      @Prop({ type: Date, default: () => new Date('2020-01-01') })
      dateProp!: Date
      @Prop({ type: Function, default: () => () => null })
      functionProp!: Function
      @Prop({ type: Symbol, default: Symbol.for('DEFAULT-SYMBOL') })
      symbolProp!: Symbol

      render() {
        return h('div')
      }
    }

    let wrapper: VueWrapper<MyComponent>

    beforeEach(() => {
      wrapper = mount(MyComponent)
    })

    it('sets prop options correctly', () => {
      expect(wrapper.vm.$options.props).toMatchSnapshot()
    })

    const properties: [keyof MyComponent, any][] = [
      ['stringProp', 'DEFAULT-STRING'],
      ['numberProp', 1234],
      ['booleanProp', true],
      ['arrayProp', ['DEFAULT-ARRAY']],
      ['objectProp', { value: 'DEFAULT-OBJECT' }],
      ['dateProp', new Date('2020-01-01')],
      ['functionProp', expect.any(Function)],
      ['symbolProp', Symbol.for('DEFAULT-SYMBOL')],
    ]

    it.each(properties)(
      `%s equals to the given prop value`,
      (property, expected) => {
        expect(wrapper.vm[property]).toEqual(expected)
      },
    )
  })
})
