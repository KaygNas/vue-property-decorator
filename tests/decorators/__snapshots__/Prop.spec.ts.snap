// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prop with constructor without prop values from its parent component sets props correctly 1`] = `
Object {
  "arrayProp": [Function],
  "booleanProp": [Function],
  "dateProp": [Function],
  "functionProp": [Function],
  "numberProp": [Function],
  "objectProp": [Function],
  "stringProp": [Function],
  "symbolProp": [Function],
}
`;

exports[`Prop with multiple constructors sets prop options correctly 1`] = `
Object {
  "myId": Array [
    [Function],
    [Function],
  ],
}
`;

exports[`Prop with prop option sets prop options correctly 1`] = `
Object {
  "arrayProp": Object {
    "0": false,
    "1": true,
    "default": [Function],
    "type": [Function],
  },
  "booleanProp": Object {
    "0": true,
    "1": true,
    "default": true,
    "type": [Function],
  },
  "dateProp": Object {
    "0": false,
    "1": true,
    "default": [Function],
    "type": [Function],
  },
  "functionProp": Object {
    "0": false,
    "1": true,
    "default": [Function],
    "type": [Function],
  },
  "numberProp": Object {
    "0": false,
    "1": true,
    "default": 1234,
    "type": [Function],
  },
  "objectProp": Object {
    "0": false,
    "1": true,
    "default": [Function],
    "type": [Function],
  },
  "stringProp": Object {
    "0": false,
    "1": true,
    "default": "DEFAULT-STRING",
    "type": [Function],
  },
  "symbolProp": Object {
    "0": false,
    "1": true,
    "default": Symbol(DEFAULT-SYMBOL),
    "type": [Function],
  },
}
`;

exports[`Prop without argument without prop values from its parent component sets props correctly 1`] = `
Object {
  "arrayProp": undefined,
  "booleanProp": undefined,
  "dateProp": undefined,
  "functionProp": undefined,
  "numberProp": undefined,
  "objectProp": undefined,
  "stringProp": undefined,
  "symbolProp": undefined,
}
`;
